util/redact/redact.go

```diff
@@ -0,0 +1,33 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redact
+
+import (
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/kv"
+)
+
+// Error is used to desensitize error message.
+func Error(err error) error {
+	if err == nil {
+		return nil
+	}
+	if !config.RedactLogEnabled() {
+		return err
+	}
+	if kv.ErrKeyExists.Equal(err) {
```

Comment:
If someone writes following code, this function could not distinguish.
```go
err = kv.ErrKeyExists
return errrors.Trace(err) // or errors.Annotate
```

Could we redact the messages at `GenWithStackByArgs`?

```go
type redactError struct {
    terror.Error
    redactPositions []int
}

func (e *redactError) GenWithStackByArgs(args ...interface{}) error {
    for _, pos := range e.redactPositions {
        args[pos] = "?"
    }
    return e.Error.GenWithStackByArgs(args)
}

func RedactError(err terror.Error, redactPositions ...int) terror.Error {
    return &redactError{err, redactPositions}
}
```

Then,
```go
var ErrKeyExists = RedactError(terror.ClassKV.New(mysql.ErrDupEntry, mysql.MySQLErrName[mysql.ErrDupEntry]), 0, 1)
```

Reply:
Great, Done.